### ISSUE TITLE
credentials_cache method takes CredentialsCache, needs to take SharedCredentialsCache

### ISSUE THREAD

**Issue by Mark-Simulacrum** [smithy-lang/smithy-rs#2945]
I *think* it's preferred to pass a credentials cache instead of a credentials provider. Right now, that's not really possible, because creating a credentials cache with credentials involves calling [CredentialsCache::create_cache](https://docs.rs/aws-credential-types/0.56.0/aws_credential_types/cache/struct.CredentialsCache.html#method.create_cache). This returns SharedCredentialsCache.

However, client builders take a CredentialsCache (not Shared) -- e.g., [ec2](https://docs.rs/aws-sdk-ec2/latest/aws_sdk_ec2/config/struct.Builder.html#method.credentials_cache), which means there doesn't seem to be a way to pass the cache-with-credentials in.

**Comment by jdisanti** [comment]
I don't think that `create_cache` is intended to be called by you during config. The expected configuration is to set both `credentials_cache` and `credentials_provider` if you want to customize caching. Something like:
```rust
let config = Config::builder()
    .credentials_cache(CredentialsCache::lazy_builder()
        // set some things ...
        .build()
    )
    .credentials_provider(my_provider)
    .build();
```
The Config builder will combine the two for you.

**Comment by Mark-Simulacrum** [comment]
I see. I think previously we were able to share a credential cache across multiple different clients, whereas with this design I think that isn't possible without some legwork to create a credential provider from a credentials cache?

**Comment by jdisanti** [comment]
Oh, sorry, I think I answered incorrectly the first time. I didn't realize the API was different between `SdkConfig` and `Config` here. Yeah, this is weird, and we should fix it.

If you can, use `SdkConfig` since that will easily share the credentials provider across clients.

**Comment by Mark-Simulacrum** [comment]
Hm, actually SdkConfig has the same problem it seems? https://docs.rs/aws-types/latest/aws_types/sdk_config/struct.Builder.html#method.credentials_cache takes a CredentialsCache.

Do you mean that we can create it via https://docs.rs/aws-config/latest/aws_config/fn.from_env.html which under the hood does the right thing?

**Comment by jdisanti** [comment]
That `CredentialsCache` that you pass into `SdkConfig` is really just a configuration that tells the builder how to create the cache rather than the cache itself. The actual cache is created inside of the `SdkConfig`'s build method, and creating the service-specific configs from that `SdkConfig` should re-use it.

**Comment by Mark-Simulacrum** [comment]
That doesn't seem to be true looking at the code? SdkConfig::build doesn't do anything with the cache ([src](https://docs.rs/aws-types/latest/src/aws_types/sdk_config.rs.html#520)), and SdkConfig::set_credentials_cache doesn't either ([src](https://docs.rs/aws-types/latest/src/aws_types/sdk_config.rs.html#328)).

The service-specific config does that: https://docs.rs/aws-sdk-ec2/latest/src/aws_sdk_ec2/config.rs.html#786-798, but that's not super helpful for my case where I'm looking to share a set of cached credentials across multiple clients with distinct configurations, including across multiple services. I suppose this is enough that within one service I can make this work (by keeping around a built Config to go back from via to_builder)...

**Comment by Mark-Simulacrum** [comment]
OK, thanks for confirming. I'll look into the workaround, it might be feasible for us to use it for the current use case.

Can you say more about anti-feature though? What are the benefits to distinct caches per service for the credentials? Is there a design doc or similar on that?

**Comment by rcoh** [comment]
this was resolved by the orchestrator however, #3427 does still exist and needs to be fixed.
